sdk/metric/sdk.go

```diff
@@ -366,14 +366,69 @@ func (m *SDK) Labels(kvs ...core.KeyValue) api.LabelSet {
 	return ls
 }
 
+// NumLabels is a part of an implementation of the export.LabelStorage
+// interface.
 func (ls *labels) NumLabels() int {
 	return ls.cachedValue.Len()
 }
 
+// GetLabel is a part of an implementation of the export.LabelStorage
+// interface.
 func (ls *labels) GetLabel(idx int) core.KeyValue {
 	return ls.cachedValue.Index(idx).Interface().(core.KeyValue)
 }
 
+// Iter is a part of an implementation of the export.Labels interface.
+func (ls *labels) Iter() export.LabelIterator {
+	return export.NewLabelIterator(ls)
+}
+
+// Encoded is a part of an implementation of the export.Labels
+// interface.
+func (ls *labels) Encoded(encoder export.LabelEncoder) string {
+	id := encoder.ID()
+	if id <= 0 {
+		// Punish misbehaving encoders by not even trying to
+		// cache them
+		return encoder.Encode(ls.Iter())
+	}
+	cachedID := atomic.LoadInt64(&ls.cachedEncoderID)
+	// If cached ID is less below zero, it means that other
+	// goroutine is currently caching the encoded labels and the
+	// ID of the encoder. Wait until it's done - it's a
+	// nonblocking op.
+	for cachedID < 0 {
+		// Let other goroutine finish its work.
+		runtime.Gosched()
+		cachedID = atomic.LoadInt64(&ls.cachedEncoderID)
+	}
+	// At this point, cachedID is either 0 (nothing cached) or
+	// some other number.
+	//
+	// If cached ID is the same as ID of the passed encoder, we've
+	// got the fast path.
+	if cachedID == id {
+		return ls.cachedEncoded
+	}
+	// If we are here, either some other encoder cached its
+	// encoded labels or the cache is still for the taking. Either
+	// way, we need to compute the encoded labels anyway.
+	encoded := encoder.Encode(ls.Iter())
+	// If some other encoder took the cache, then we just return
```

Comment:
This caching scheme seems to only cache for a single encoder. Specifically the first encoder called. Any reason we cannot use a `sync.Map` to cache for all the encoders?

Reply:
Yeah. I don't know if we currently have a use case of having different label encoder for batcher and different for exporter. For now we expect them to use the same encoder, so caching scheme for a single encoder makes sense.

Reply:
It's OK to support one for now, since that's the status quo. In the PR supporting OTEP 78 I let the labels object support caching up to three encodings. We could do that later.